### ISSUE TITLE
Asan: detected memory leaks, server can not startup

### ISSUE THREAD

**Issue by iamfaith** [ossrs/srs#3238]
> Note: Please read FAQ before file an issue, see 2716

> Note: Before asking a question, please refer to the FAQ, specifically 2716.

**Description (描述)**

1. git clone -b develop https://gitee.com/ossrs/srs.git &&
cd srs/trunk && ./configure && make && ./objs/srs -c conf/srs.conf
2. Error message:
```
[2022-11-15 09:23:01.118][INFO][765537][1958j687] XCORE-SRS/5.0.86(Bee)
[2022-11-15 09:23:01.118][INFO][765537][1958j687] config parse complete
[2022-11-15 09:23:01.118][INFO][765537][1958j687] you can check log by: tail -n 30 -f ./objs/srs.log
[2022-11-15 09:23:01.118][INFO][765537][1958j687] please check SRS by: ./etc/init.d/srs status

=================================================================
==765539==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 648 byte(s) in 9 object(s) allocated from:
    #0 0x7f2126ad1587 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cc:104
    #1 0x56517a91e7fb in discover_network_iface(ifaddrs*, std::vector<SrsIPAddress*, std::allocator<SrsIPAddress*> >&, std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >&, bool, bool) src/protocol/srs_protocol_utility.cpp:684
    #2 0x56517a91f883 in retrieve_local_ips() src/protocol/srs_protocol_utility.cpp:766
    #3 0x56517a92004b in srs_get_local_ips() src/protocol/srs_protocol_utility.cpp:802
    #4 0x56517aaa8078 in SrsConfig::check_normal_config() src/app/srs_app_config.cpp:2410
    #5 0x56517aaa525b in SrsConfig::check_config() src/app/srs_app_config.cpp:2236
    #6 0x56517ae0f402 in do_main(int, char**, char**) src/main/srs_main_server.cpp:205
    #7 0x56517ae0f770 in main src/main/srs_main_server.cpp:231
    #8 0x7f2126473082 in __libc_start_main ../csu/libc-start.c:308

Direct leak of 144 byte(s) in 2 object(s) allocated from:
    #0 0x7f2126ad1587 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cc:104
    #1 0x56517a91e7fb in discover_network_iface(ifaddrs*, std::vector<SrsIPAddress*, std::allocator<SrsIPAddress*> >&, std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >&, bool, bool) src/protocol/srs_protocol_utility.cpp:684
    #2 0x56517a91f581 in retrieve_local_ips() src/protocol/srs_protocol_utility.cpp:745
    #3 0x56517a92004b in srs_get_local_ips() src/protocol/srs_protocol_utility.cpp:802
    #4 0x56517aaa8078 in SrsConfig::check_normal_config() src/app/srs_app_config.cpp:2410
    #5 0x56517aaa525b in SrsConfig::check_config() src/app/srs_app_config.cpp:2236
    #6 0x56517ae0f402 in do_main(int, char**, char**) src/main/srs_main_server.cpp:205
    #7 0x56517ae0f770 in main src/main/srs_main_server.cpp:231
    #8 0x7f2126473082 in __libc_start_main ../csu/libc-start.c:308

Indirect leak of 345 byte(s) in 9 object(s) allocated from:
    #0 0x7f2126ad1587 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cc:104
    #1 0x7f21268f029e in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_assign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/lib/x86_64-linux-gnu/libstdc++.so.6+0x14329e)

SUMMARY: AddressSanitizer: 1137 byte(s) leaked in 20 allocation(s).

=================================================================
==765538==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 648 byte(s) in 9 object(s) allocated from:
    #0 0x7f2126ad1587 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cc:104
    #1 0x56517a91e7fb in discover_network_iface(ifaddrs*, std::vector<SrsIPAddress*, std::allocator<SrsIPAddress*> >&, std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >&, bool, bool) src/protocol/srs_protocol_utility.cpp:684
    #2 0x56517a91f883 in retrieve_local_ips() src/protocol/srs_protocol_utility.cpp:766
    #3 0x56517a92004b in srs_get_local_ips() src/protocol/srs_protocol_utility.cpp:802
    #4 0x56517aaa8078 in SrsConfig::check_normal_config() src/app/srs_app_config.cpp:2410
    #5 0x56517aaa525b in SrsConfig::check_config() src/app/srs_app_config.cpp:2236
    #6 0x56517ae0f402 in do_main(int, char**, char**) src/main/srs_main_server.cpp:205
    #7 0x56517ae0f770 in main src/main/srs_main_server.cpp:231
    #8 0x7f2126473082 in __libc_start_main ../csu/libc-start.c:308

Direct leak of 144 byte(s) in 2 object(s) allocated from:
    #0 0x7f2126ad1587 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cc:104
    #1 0x56517a91e7fb in discover_network_iface(ifaddrs*, std::vector<SrsIPAddress*, std::allocator<SrsIPAddress*> >&, std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >&, bool, bool) src/protocol/srs_protocol_utility.cpp:684
    #2 0x56517a91f581 in retrieve_local_ips() src/protocol/srs_protocol_utility.cpp:745
    #3 0x56517a92004b in srs_get_local_ips() src/protocol/srs_protocol_utility.cpp:802
    #4 0x56517aaa8078 in SrsConfig::check_normal_config() src/app/srs_app_config.cpp:2410
    #5 0x56517aaa525b in SrsConfig::check_config() src/app/srs_app_config.cpp:2236
    #6 0x56517ae0f402 in do_main(int, char**, char**) src/main/srs_main_server.cpp:205
    #7 0x56517ae0f770 in main src/main/srs_main_server.cpp:231
    #8 0x7f2126473082 in __libc_start_main ../csu/libc-start.c:308

Indirect leak of 345 byte(s) in 9 object(s) allocated from:
    #0 0x7f2126ad1587 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cc:104
    #1 0x7f21268f029e in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_assign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/lib/x86_64-linux-gnu/libstdc++.so.6+0x14329e)

SUMMARY: AddressSanitizer: 1137 byte(s) leaked in 20 allocation(s).

=================================================================
==765537==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 648 byte(s) in 9 object(s) allocated from:
    #0 0x7f2126ad1587 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cc:104
    #1 0x56517a91e7fb in discover_network_iface(ifaddrs*, std::vector<SrsIPAddress*, std::allocator<SrsIPAddress*> >&, std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >&, bool, bool) src/protocol/srs_protocol_utility.cpp:684
    #2 0x56517a91f883 in retrieve_local_ips() src/protocol/srs_protocol_utility.cpp:766
    #3 0x56517a92004b in srs_get_local_ips() src/protocol/srs_protocol_utility.cpp:802
    #4 0x56517aaa8078 in SrsConfig::check_normal_config() src/app/srs_app_config.cpp:2410
    #5 0x56517aaa525b in SrsConfig::check_config() src/app/srs_app_config.cpp:2236
    #6 0x56517ae0f402 in do_main(int, char**, char**) src/main/srs_main_server.cpp:205
    #7 0x56517ae0f770 in main src/main/srs_main_server.cpp:231
    #8 0x7f2126473082 in __libc_start_main ../csu/libc-start.c:308

Direct leak of 144 byte(s) in 2 object(s) allocated from:
    #0 0x7f2126ad1587 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cc:104
    #1 0x56517a91e7fb in discover_network_iface(ifaddrs*, std::vector<SrsIPAddress*, std::allocator<SrsIPAddress*> >&, std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >&, bool, bool) src/protocol/srs_protocol_utility.cpp:684
    #2 0x56517a91f581 in retrieve_local_ips() src/protocol/srs_protocol_utility.cpp:745
    #3 0x56517a92004b in srs_get_local_ips() src/protocol/srs_protocol_utility.cpp:802
    #4 0x56517aaa8078 in SrsConfig::check_normal_config() src/app/srs_app_config.cpp:2410
    #5 0x56517aaa525b in SrsConfig::check_config() src/app/srs_app_config.cpp:2236
    #6 0x56517ae0f402 in do_main(int, char**, char**) src/main/srs_main_server.cpp:205
    #7 0x56517ae0f770 in main src/main/srs_main_server.cpp:231
    #8 0x7f2126473082 in __libc_start_main ../csu/libc-start.c:308

Indirect leak of 345 byte(s) in 9 object(s) allocated from:
    #0 0x7f2126ad1587 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cc:104
    #1 0x7f21268f029e in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_assign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/lib/x86_64-linux-gnu/libstdc++.so.6+0x14329e)

SUMMARY: AddressSanitizer: 1137 byte(s) leaked in 20 allocation(s).
```

> Please describe your issue here (描述你遇到了什么问题)

The server can not startup, and I use the default configuration.

**Expect (Expected Behavior)**

> Please describe your expectation (Describe what you expect to happen)

Should work as expected

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
What's the log?

**Comment by winlinvip** [comment]
What's the version?

**Comment by winlinvip** [comment]
It starts ok, in daemon mode, please read wiki https://ossrs.io/lts/en-us/docs/v5/doc/getting-started-build

Check SRS by http://localhost:8080/ or:

```bash
# Check the process status
./etc/init.d/srs status

# Check the SRS logs
tail -n 30 -f ./objs/srs.log
```

However, we must fix the confusing logs.

**Comment by winlinvip** [comment]
Also confusing logs when stopping SRS: 

```bash
$ ./etc/init.d/srs stop

Stopping SRS(pid 228923)...                                [  OK  ]

=================================================================
==228923==ERROR: LeakSanitizer: detected memory leaks


Direct leak of 216 byte(s) in 3 object(s) allocated from:
    #0 0x7f9d9e09e587 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cc:104
    #1 0x560e65d43ca0 in discover_network_iface(ifaddrs*, std::vector<SrsIPAddress*, std::allocator<SrsIPAddress*> >&, std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >&, bool, bool) src/protocol/srs_protocol_utility.cpp:695
    #2 0x560e65d44c72 in retrieve_local_ips() src/protocol/srs_protocol_utility.cpp:773
    #3 0x560e65d4543a in srs_get_local_ips() src/protocol/srs_protocol_utility.cpp:809
    #4 0x560e65ecd862 in SrsConfig::check_normal_config() src/app/srs_app_config.cpp:2410
    #5 0x560e65ecaa45 in SrsConfig::check_config() src/app/srs_app_config.cpp:2236
    #6 0x560e66234bec in do_main(int, char**, char**) src/main/srs_main_server.cpp:205
    #7 0x560e66234f5a in main src/main/srs_main_server.cpp:231
    #8 0x7f9d9da40082 in __libc_start_main ../csu/libc-start.c:308

Direct leak of 72 byte(s) in 1 object(s) allocated from:
    #0 0x7f9d9e09e587 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cc:104
    #1 0x560e65d43ca0 in discover_network_iface(ifaddrs*, std::vector<SrsIPAddress*, std::allocator<SrsIPAddress*> >&, std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >&, bool, bool) src/protocol/srs_protocol_utility.cpp:695
    #2 0x560e65d44970 in retrieve_local_ips() src/protocol/srs_protocol_utility.cpp:752
    #3 0x560e65d4543a in srs_get_local_ips() src/protocol/srs_protocol_utility.cpp:809
    #4 0x560e65ecd862 in SrsConfig::check_normal_config() src/app/srs_app_config.cpp:2410
    #5 0x560e65ecaa45 in SrsConfig::check_config() src/app/srs_app_config.cpp:2236
    #6 0x560e66234bec in do_main(int, char**, char**) src/main/srs_main_server.cpp:205
    #7 0x560e66234f5a in main src/main/srs_main_server.cpp:231
    #8 0x7f9d9da40082 in __libc_start_main ../csu/libc-start.c:308

Direct leak of 40 byte(s) in 1 object(s) allocated from:
    #0 0x7f9d9e09ca06 in __interceptor_calloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:153
    #1 0x560e6623e337 in _st_epoll_init /media/psf/Home/git/srs/trunk/objs/Platform-Linux-5.15.0-GCC9.4.0-SRS5-x86_64/st-srs/event.c:866
    #2 0x560e66239439 in st_init /media/psf/Home/git/srs/trunk/objs/Platform-Linux-5.15.0-GCC9.4.0-SRS5-x86_64/st-srs/sched.c:197
    #3 0x560e65da79c6 in srs_st_init() src/protocol/srs_protocol_st.cpp:62
    #4 0x560e660fe2e6 in SrsThreadPool::setup_thread_locals() src/app/srs_app_threads.cpp:556
    #5 0x560e66100337 in SrsThreadPool::start(void*) src/app/srs_app_threads.cpp:771
    #6 0x7f9d9df6e608 in start_thread /build/glibc-SzIz7B/glibc-2.31/nptl/pthread_create.c:477

Indirect leak of 65536 byte(s) in 1 object(s) allocated from:
    #0 0x7f9d9e09ca06 in __interceptor_calloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:153
    #1 0x560e6623e41f in _st_epoll_init /media/psf/Home/git/srs/trunk/objs/Platform-Linux-5.15.0-GCC9.4.0-SRS5-x86_64/st-srs/event.c:882
    #2 0x560e66239439 in st_init /media/psf/Home/git/srs/trunk/objs/Platform-Linux-5.15.0-GCC9.4.0-SRS5-x86_64/st-srs/sched.c:197
    #3 0x560e65da79c6 in srs_st_init() src/protocol/srs_protocol_st.cpp:62
    #4 0x560e660fe2e6 in SrsThreadPool::setup_thread_locals() src/app/srs_app_threads.cpp:556
    #5 0x560e66100337 in SrsThreadPool::start(void*) src/app/srs_app_threads.cpp:771
    #6 0x7f9d9df6e608 in start_thread /build/glibc-SzIz7B/glibc-2.31/nptl/pthread_create.c:477

Indirect leak of 49152 byte(s) in 1 object(s) allocated from:
    #0 0x7f9d9e09c808 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:144
    #1 0x560e6623e48b in _st_epoll_init /media/psf/Home/git/srs/trunk/objs/Platform-Linux-5.15.0-GCC9.4.0-SRS5-x86_64/st-srs/event.c:891
    #2 0x560e66239439 in st_init /media/psf/Home/git/srs/trunk/objs/Platform-Linux-5.15.0-GCC9.4.0-SRS5-x86_64/st-srs/sched.c:197
    #3 0x560e65da79c6 in srs_st_init() src/protocol/srs_protocol_st.cpp:62
    #4 0x560e660fe2e6 in SrsThreadPool::setup_thread_locals() src/app/srs_app_threads.cpp:556
    #5 0x560e66100337 in SrsThreadPool::start(void*) src/app/srs_app_threads.cpp:771
    #6 0x7f9d9df6e608 in start_thread /build/glibc-SzIz7B/glibc-2.31/nptl/pthread_create.c:477

Indirect leak of 106 byte(s) in 3 object(s) allocated from:
    #0 0x7f9d9e09e587 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cc:104
    #1 0x7f9d9debd29e in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_assign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/lib/x86_64-linux-gnu/libstdc++.so.6+0x14329e)

SUMMARY: AddressSanitizer: 115122 byte(s) leaked in 10 allocation(s).
SRS stopped by SIGTERM                                     [  OK  ]
```

**Comment by winlinvip** [comment]
It also returns none 0, which is error, we will fix this.

```bash
# ./objs/srs -c conf/srs.conf
# echo $?
1

# ./etc/init.d/srs  status
SRS(pid 241539) is running.                                [  OK  ]
# echo $?
0
```

**Comment by winlinvip** [comment]
Fixed in 5.0.91
